File: plugin/src/main/java/com/jtprince/bingo/plugin/BingoPlayerTeam.java
package com.jtprince.bingo.plugin;

import org.bukkit.Bukkit;
import org.bukkit.OfflinePlayer;
import org.bukkit.entity.Player;

import java.util.ArrayList;
import java.util.Collection;
import java.util.UUID;
import java.util.stream.Collectors;

public class BingoPlayerTeam implements BingoPlayer {
    private final Collection<UUID> playerUuids;
    private final String teamName;

    public BingoPlayerTeam(String teamName, Collection<OfflinePlayer> teammates) {
        this.playerUuids = teammates.stream().map(OfflinePlayer::getUniqueId)
            .collect(Collectors.toCollection(ArrayList::new));
        this.teamName = teamName;
    }

    @Override
    public String getName() {
        return this.teamName;
    }

    @Override
    public Collection<Player> getBukkitPlayers() {
        return this.playerUuids.stream().map(Bukkit::getPlayer)
            .collect(Collectors.toCollection(ArrayList::new));
    }
}

File: kotlinplugin/src/main/kotlin/com/jtprince/bingo/kplugin/player/BingoPlayerTeam.kt
package com.jtprince.bingo.kplugin.player

import net.kyori.adventure.text.Component
import net.kyori.adventure.text.serializer.plain.PlainComponentSerializer
import org.bukkit.Bukkit
import org.bukkit.OfflinePlayer
import org.bukkit.entity.Player
import java.util.*

/**
 * A team of 1 or more players that will share a win condition and world.
 */
class BingoPlayerTeam(private val teamName: Component,
                      teammates: Collection<OfflinePlayer>) : BingoPlayer() {
    private val playerUuids: Collection<UUID> = teammates.map(OfflinePlayer::getUniqueId)

    override val name: String
        get() = PlainComponentSerializer.plain().serialize(teamName)

    override val formattedName: Component
        get() = teamName

    override val bukkitPlayers: Collection<Player>
        get() = playerUuids.mapNotNull(Bukkit::getPlayer).filter(OfflinePlayer::isOnline)

    override val offlinePlayers: Collection<OfflinePlayer>
        get() = playerUuids.map(Bukkit::getOfflinePlayer)
}
